src/server/transaction.cc

```diff
@@ -711,30 +692,49 @@ void Transaction::ScheduleInternal() {
   DCHECK_GT(unique_shard_cnt_, 0u);
   DCHECK(!IsAtomicMulti() || cid_->IsMultiTransactional());
 
-  DVLOG(1) << "ScheduleInternal " << cid_->name() << " on " << unique_shard_cnt_ << " shards";
+  // Try running immediately (during scheduling) if we're concluding and either:
+  // - have a single shard, and thus never have to cancel scheduling due to reordering
+  // - run as an idempotent command, meaning we can safely repeat the operation if scheduling fails
+  bool can_run_immediately = !IsGlobal() && (coordinator_state_ & COORD_CONCLUDING) &&
+                             (unique_shard_cnt_ == 1 || (cid_->opt_mask() & CO::IDEMPOTENT));
+
+  DVLOG(1) << "ScheduleInternal " << cid_->name() << " on " << unique_shard_cnt_ << " shards "
+           << " immediate run: " << can_run_immediately;
 
   auto is_active = [this](uint32_t i) { return IsActive(i); };
 
   // Loop until successfully scheduled in all shards.
   while (true) {
     stats_.schedule_attempts++;
 
-    txid_ = op_seq.fetch_add(1, memory_order_relaxed);
+    // This is a contention point for all threads - avoid using it unless necessary.
+    // Single shard operations can assign txid later if the immediate run failed.
+    if (unique_shard_cnt_ > 1)
+      txid_ = op_seq.fetch_add(1, memory_order_relaxed);
+
     InitTxTime();
 
     atomic_uint32_t schedule_fails = 0;
-    auto cb = [this, &schedule_fails](EngineShard* shard) {
-      if (!ScheduleInShard(shard)) {
+    auto cb = [this, &schedule_fails, can_run_immediately]() {
+      if (!ScheduleInShard(EngineShard::tlocal(), can_run_immediately)) {
         schedule_fails.fetch_add(1, memory_order_relaxed);
       }
+      run_barrier_.Dec();
     };
-    shard_set->RunBriefInParallel(std::move(cb), is_active);
+
+    run_barrier_.Start(unique_shard_cnt_);
+    if (CanRunInlined()) {
```

Comment:
why not?

```
if (CanRunInlined()) {
  CHECK(ScheduleInShard(EngineShard::tlocal(), can_run_immediately));
} else {
  auto cb = ..
  run_barrier_.Start(unique_shard_cnt_);
  IterateActiveShards([cb](const auto& sd, ShardId i) { shard_set->Add(i, cb); });
  run_barrier_.Wait();
}
```

Reply:
True, good idea... I just thought you don't like such checks 😄

Reply:
It fails our DCHECK logic for the barrier being active, I changed it a little to make it fit